### PR TITLE
[libc] Allow each function can have extra attributes by defining LLVM_LIBC_FUNCTION_ATTR_func macro.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/src/__support/common.h [hunk 1]
```diff
@@ -21,9 +21,24 @@
 #define LLVM_LIBC_FUNCTION_ATTR
 #endif
 
+// Allow each function `func` can have extra attributes specified by defining:
+// `LLVM_LIBC_FUNCTION_ATTR_func` macro, which should always start with
+// "LLVM_LIBC_EMPTY,"
+//
+// For example:
+// #define LLVM_LIBC_FUNCTION_ATTR_memcpy LLVM_LIBC_EMPTY, __attribute__((weak))
```
[the remark this answers is not in the document]
Comment (nickdesaulniers) [reply libc/src/__support/common.h:29]: Perhaps worth documenting any new command line options in docs/dev/implementation_standard.rst.

File: libc/src/__support/common.h [hunk 2]
```diff
@@ -21,9 +21,24 @@
...
+#define LLVM_LIBC_EMPTY
+
+#define GET_SECOND(first, second, ...) second
+#define EXPAND_THEN_SECOND(name) GET_SECOND(name, LLVM_LIBC_EMPTY, )
+
+#define LLVM_LIBC_ATTR(name) EXPAND_THEN_SECOND(LLVM_LIBC_FUNCTION_ATTR_##name)
+#define EXPAND_ATTR(name) LLVM_LIBC_ATTR(name)
```
Comment (vonosmas) [inline libc/src/__support/common.h:36]: I don't think you need `EXPAND_ATTR` and can just use `LLVM_LIBC_ATTR` below?

File: libc/src/__support/common.h (hunk 1 above)
Comment (vonosmas) [inline libc/src/__support/common.h:29]: Should this be `LLVM_LIBC_EMPTY,__attribute__((weak))` (w/o space)?

If not, use `"LLVM_LIBC_EMPTY, "` on line 26 above.

Comment (lntue) [reply libc/src/__support/common.h:29]: Updated line 26.  Passing the macro from `EXPAND_THEN_SECOND` to `SECOND`  will ignore the spaces there, give some extra spaces should make it better for readability.

File: libc/src/__support/common.h (hunk 2 above)
Comment (lntue) [reply libc/src/__support/common.h:36]: I think you're right, `EXPAND_ATTR` is not needed.

File: libc/src/__support/common.h [hunk 3]
```diff
@@ -21,9 +21,24 @@
 #define LLVM_LIBC_FUNCTION_ATTR
 #endif
 
+// Allow each function `func` can have extra attributes specified by defining:
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/__support/common.h:24]: Done.

File: utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl [hunk 4]
```diff
@@ -59,7 +59,7 @@ def libc_function(
         srcs,
         weak = False,
         copts = None,
-        local_defines = None,
+        local_defines = [],
```
Comment (vonosmas) [inline utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl:62]: I would propose to remove this argument altogether - I don't think it has uses now, and we don't want it to be used in the future - and favor the more uniform build of all `libc_function`s instead.

File: utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl [hunk 5]
```diff
@@ -108,16 +108,23 @@ def libc_function(
...
     )
 
     # This second target is the llvm libc C function with either a default or hidden visibility.
     # All other functions are hidden.
-    func_attrs = ["__attribute__((visibility(\"default\")))"]
-    if weak:
-        func_attrs = func_attrs + ["__attribute__((weak))"]
-    local_defines = local_defines or ["LIBC_COPT_PUBLIC_PACKAGING"]
-    local_defines = local_defines + ["LLVM_LIBC_FUNCTION_ATTR='%s'" % " ".join(func_attrs)]
+    global_func_attrs = [
```
Comment (vonosmas) [inline utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl:117]: Remove temp variable for clarity - you can just use `LLVM_LIBC_FUNCTION_ATTR='[[gnu:visibility(\"default\")]]'` below instead.

**Review by vonosmas — APPROVED** [review]
LGTM, but please get confirmation from @nickdesaulniers  as well.

File: libc/src/__support/common.h [hunk 6]
```diff
@@ -21,9 +21,25 @@
...
+// "LLVM_LIBC_EMPTY, "
+//
+// For examples:
+// #define LLVM_LIBC_FUNCTION_ATTR_memcpy LLVM_LIBC_EMPTY, [[gnu::weak]]
+// #define LLVM_LIBC_FUNCTION_ATTR_memchr LLVM_LIBC_EMPTY, [[gnu::weak]]       \
+//                                        [[gnu::visibility("default")]]
+#define LLVM_LIBC_EMPTY
+
+#define GET_SECOND(first, second, ...) second
+#define EXPAND_THEN_SECOND(name) GET_SECOND(name, LLVM_LIBC_EMPTY, )
```
Comment (nickdesaulniers) [inline libc/src/__support/common.h:35]: I think this still works: https://godbolt.org/z/dT4xTYqoY
```suggestion
#define EXPAND_THEN_SECOND(name) GET_SECOND(name, LLVM_LIBC_EMPTY)
```

**Review by nickdesaulniers — APPROVED** [review]

File: libc/src/__support/common.h (hunk 6 above)
Comment (lntue) [reply libc/src/__support/common.h:35]: Done.

File: utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl (hunk 5 above)
Comment (lntue) [reply utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl:117]: Done.

File: utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl (hunk 4 above)
Comment (lntue) [reply utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl:62]: Let me just keep it here for now in this PR, and leave the clean up work of bazel build for later when we refactor the bazel builds.

**Pull request merged**